File: kontringsattack.cpp

#include <iostream>
#include <algorithm>
#include <vector>

int main() {
    
    
    int N;
    std::cin >> N;

    std::vector<int> diffs(N);
    for (int i = 0; i < N; i++) {
        int F, S;
        std::cin >> F >> S;
        diffs[i] = F - S;
    }

    std::vector<int> unique_diffs = {};
    for (int i = 0; i < N; i++) {
        int dif = abs(diffs[i]);
        if (std::find(unique_diffs.begin(), unique_diffs.end(), dif) == unique_diffs.end())
            unique_diffs.push_back(dif);
    }

    std::sort(unique_diffs.begin(), unique_diffs.end());
    std::sort(diffs.begin(), diffs.end(), [](int n, int m) {return (abs(n) < abs(m));});

    int F_wins = std::count_if(diffs.begin(), diffs.end(), [](int d) {return (d > 0);});
    int S_wins = std::count_if(diffs.begin(), diffs.end(), [](int d) {return (d < 0);});

    int match_dif = F_wins - S_wins;

    int highest_match_dif = match_dif;
    int best_bound = 0;

    int piv = 0;

    for (int i = 0; i < unique_diffs.size(); i++) {

        for (piv; piv < N; piv++) {

            if (abs(diffs[piv]) != unique_diffs[i])
                break;
            
            if (diffs[piv] > 0)
                match_dif -= 1;
            else
                match_dif += 1;

        }

        if (match_dif > highest_match_dif) {
            highest_match_dif = match_dif;
            best_bound = unique_diffs[i];
        }

    }

    std::cout << best_bound << std::endl;

}

File: kontringsattack.py

N = int(input())
matches = [list(map(int, input().split())) for _ in range(N)]

diffs = list(map(lambda p: p[0] - p[1], matches))
diffs.sort(key=abs)

unique_diffs = []
for diff in diffs:
    if abs(diff) not in unique_diffs:
        unique_diffs.append(abs(diff))

unique_diffs.sort()

F_wins = sum([1 if d > 0 else 0 for d in diffs])
S_wins = sum([1 if d < 0 else 0 for d in diffs])

match_dif = F_wins - S_wins

best_match_dif = F_wins - S_wins
best_bound = 0

piv = 0

for unique_diff in unique_diffs:
    
    # går att förbättra så att man bara går igenom diffs en gång istället för len(unique_diffs) gånger (max N gånger)

    for i in range(piv, N):
        piv = i
        if abs(diffs[i]) != unique_diff:
            break
        if diffs[piv] > 0:
            match_dif -= 1
        elif diffs[piv] < 0:
            match_dif += 1

    if match_dif > best_match_dif:
        best_match_dif = match_dif
        best_bound = unique_diff

print(best_bound)

